### DISCUSSION TITLE
Avoid re-running wokflow for the same commit

### DISCUSSION THREAD

**Discussion by amezin** [community/community#27031] [Actions]
<p>With <code>on: push</code> a workflow sometimes runs multiple times for the same commit (pushed as a branch and then a tag, or fast-forward merge). Is there a way to avoid this (for one specific workflow)? I. e. don’t run the workflow if it already ran for the same commit id.</p>

**Comment by shrink** [comment, marked as answer]
<p>Yes, you can use the <a href="https://docs.github.com/en/rest/reference/checks" rel="noopener nofollow ugc">Check Runs API</a> to identify <a href="https://docs.github.com/en/rest/reference/checks#list-check-runs-for-a-git-reference" rel="noopener nofollow ugc">Workflow runs for a ref</a>. As luck would have it, someone has already built a comprehensive Action for this use-case: <a href="https://github.com/marketplace/actions/skip-duplicate-actions" rel="noopener nofollow ugc">fkirc/skip-duplicate-actions</a>.</p>
<ol>
<li>Add a new “pre” job to your Workflow, this job uses <code>fkirc/skip-duplicate-actions</code> to determine if your main job should be skipped</li>
<li>Add a condition to your main job using the <code>should_skip</code> output of the “pre” job.</li>
</ol>
<p>For example, adapted from the <code>fkirc/skip-duplicate-actions</code> readme:</p>
<pre><code class="lang-yaml">jobs:
  pre_job:
    runs-on: ubuntu-latest
    outputs:
      should_skip: ${{ steps.skip_check.outputs.should_skip }}
    steps:
      - id: skip_check
        uses: fkirc/skip-duplicate-actions@v3.4.0
        with:
          skip_after_successful_duplicate: 'true'
  main_job:
    needs: pre_job
    if: ${{ needs.pre_job.outputs.should_skip != 'true' }}
    runs-on: ubuntu-latest
    steps:
      - run: echo "Running slow tests..." &amp;&amp; sleep 30
</code></pre>

**Comment by jsoref** [comment]
<p>I use:</p>
<pre><code class="lang-auto">on:
  push:
    branches:
      - "**"
    tags-ignore:
      - "**"
</code></pre>
<p>Technically if someone pushed a second branch for the same commit, it’d run, but mostly it guards against tags triggering a thing again.</p>

**Comment by kmturley** [comment]
I think what's confusing is that you want to show the result of tests in your PR.

So you add the `pull_request` event, but that creates the duplicate run:
```
name: Test
on:
  push:
  pull_request:
```
Results in:
<img width="810" height="267" alt="Screenshot 2025-12-02 at 9 09 43 PM" src="https://github.com/user-attachments/assets/ca1d6739-51a2-4089-a5b2-9975bb1ea164" />


However if you remove `pull_request` event you still see the Test results in the PR, without duplicates:
```
name: Test
on:
  push:
```
Results in:
<img width="916" height="370" alt="Screenshot 2025-12-02 at 9 09 17 PM" src="https://github.com/user-attachments/assets/e9df889f-c6da-4bbd-a5db-3cb367ba9f88" />

**Reply by jsoref** [reply]
There isn't much you can do about that (essentially it's a training issue).
For [check-spelling](https://github.com/check-spelling/check-spelling), I implemented [suppress_push_for_open_pull_request](https://docs.check-spelling.dev/Configuration#suppresspushforopenpullrequest) which does some magic to identify if there's an open PR in the same repository and will skip doing the work (which may not be proper if there's a live PR where the merge product is what matters).

Also, as the minimum billing unit is 1m, there's _really_ very little you can/show do about it.

**Reply by kmturley** [reply]
I use this in my repos:
```
on:
  push:
    branches-ignore:
      - main
    tags-ignore:
      - 'v*'
```

Which only runs the workflow once on `git push` which is usually a group of commits.
Then it is never run again, even if a PR is opened, merged or a tag is pushed
